- document deployment approaches (#1431)
conan-io/docs: integrations/deployment.rst

.. _deployment:

How to deploy C++ applications with conan
=========================================

> For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 

Reviewer: ```suggestion
For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 
```
Irrelevant introduction

Author: removed

> and how to generate a deployable artifact after the build.

Reviewer: ```suggestion

```

> Consider we have a C++ application, with a complex dependency graph, and use conan to manage these dependencies from time to time, 

Reviewer: ```suggestion

```

> it may be possible to significantly reduce the complexity of deployment via the help of some built-in conan facilities. 

Reviewer: ```suggestion

```

> Once you have your project built with Conan packages you can extract the needed artifacts to a local directory. Check out 

Reviewer: ```suggestion
If you have a project with all the dependencies managed with Conan and you want to deploy it the process is:

- Extract the needed artifacts to a local directory. Check out the section :ref:`using_deploy_generator` and :ref:`using_json_generator` in order to get information on how to extract build artifacts.  
- Convert the artifacts (typically executables, shared libraries and assets) to a different deploy format. In the following sections, we are going to review some of the deploy technologies.

```

- Define new sub-setting, for instance `os.distro`, as explained in the section :ref:`add_new_sub_settings`.
- Define compatibility mode, as explained by sections :ref:`method_package_id` and :ref:`method_build_id` (e.g. you may consider some ``Ubuntu`` and ``Debian`` packages to be compatible with each other)
- Generate different packages for each distribution.
- Generate deployable artifacts for each distribution.

C++ standard library
++++++++++++++++++++

Usually, the default C++ standard library is `libstdc++ <https://gcc.gnu.org/onlinedocs/libstdc++/>`_, but
`libc++ <https://libcxx.llvm.org/>`_ and `stlport <http://www.stlport.org/>`_ are other well-known implementations.

Similarly to the standard C library `glibc`, running the application linked with libstdc++ in the older system may result in an error:

.. code-block:: console

    $ /hello
    /hello: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /hello)
    /hello: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /hello)

Fortunately, this is much easier to address by just adding ``-static-libstdc++`` compiler flag. Unlike C runtime, C++ runtime can be 

> linked statically safely, because it doesn't use system calls directly, but instead relies on ``libc`` to provide required wrappers.

Reviewer 2: 👍 much better
